### ISSUE TITLE
Generate one fixup per commit, not per-hunk?

### ISSUE THREAD

**Issue by nickolay** [tummychow/git-absorb#19]
Currently running git-absorb with non-trivial changes creates LOTS of fixup commits with the same message, because a separate fixup commit is created for each hunk. Is there a reason to do this or could we squash those into a few commits?

**Comment by tummychow** [comment]
as you've observed, the user gets no benefit from having one fixup commit per hunk if they're both absorbing into the same commit. however, it makes the absorption loop a LOT easier to write. i would probably accept a PR for this if it could be implemented in a legible way, but instinctively i feel like this would be a scary refactor to write

**Comment by stof** [comment]
the advantage of limiting the number of fixup commits appears when a human inspects them.
For instance, in our company, we regularly push our fixup commits into our feature branches on github, so that the reviewer of the PR can easily see what has changed since last time, and we squash them only at the end before merging the PR.

**Comment by tummychow** [comment]
> 1. figure out all the commits to make (same as now)
> 2. merge those with same fixup target together

yes, i think this is the most likely implementation that would work. you have to wait to make commits until after the entire absorption loop is done.

> 1. commit the fixups
> 2. go back and run a rebase which squashes the fixups among themselves, but not with the original target.

where as this approach sounds quite fiddly to me - because you're performing two interactive rebases in a row (one to combine fixups together, and then another to absorb the fixups into the actual commits) and you have to keep the commit messages organized.

**Comment by craftey** [comment]
I have tested the new option `--one-fixup-per-commit`.  Within my first test it seemed not to work as expected:

I have a feature with one commit that contains two hunks.
Now I change some bits in both hunks, stage the changes and run `git absorb --one-fixup-per-commit`.
This creates only one fixup commit for the first hunk. The second change is kept in git stage.
Running `git absorb --one-fixup-per-commit` again will create another fixup commit for the left change.
I would have expected one fixup commit with both hunks from the first run.

If I run just `git absorb` I get two fixup commits with same title, ie one per hunk.

**Comment by tummychow** [comment]
@craftey ugh, yes i can repro. it's some off-by-one line offset issue. i'm not going to revert the pr but the feature will stay out of a release until i (or someone else) has time to track it down.

**Comment by craftey** [comment]
Yes, maybe @rbartlensky can have a look.

**Comment by rbartlensky** [comment]
@craftey @tummychow I can have a look this week -- thanks for the repro case!

**Comment by rbartlensky** [comment]
Opened #93 to fix the issue

**Comment by nickolay** [comment]
The feature is available in the 0.6.11 version on an opt-in basis:

    -F, --one-fixup-per-commit    Only generate one fixup per commit

I've been happily using it since August by adding it to my sublime merge configuration

    $ cat /Users/nickolay/Library/Application\ Support/Sublime\ Merge/Packages/User/Default.sublime-commands
    [
        {
            "caption": "Git absorb",
            "command": "git",
            "args": {"argv": ["absorb", "-F"]}
        }
    ]

As far as I am concerned the only reason to keep this issue open is to maybe make it the default, on which Stephen [commented elsewhere](https://github.com/tummychow/git-absorb/pull/91#issuecomment-1685455294):

> i want to let it bake for a while, given the amount of code getting moved around. i would accept a pr to add a git config option for it though

Huge thanks again to everyone who made this happen — @rbartlensky, @tummychow, @craftey!

**Comment by kiprasmel** [comment]
fyi you can now enable `-F` to always be true via the new config since #101:

```ini
[absorb]
    oneFixupPerCommit = true
```

**Comment by blairconrad** [comment]
Is this issue being left open with the intention of making one-fixup-per-commit the default, as @nickolay [suggested](https://github.com/tummychow/git-absorb/issues/19#issuecomment-1872643617)?

If so, is there some condition that is preventing the change from being made? Accumulations of other changes before a "major release" or similar?

If the behaviour were to be made default, would you 

* make a hard switch, eliminating the existing CLI flag (and configuration option)
* leave the existing flag, perhaps warning when it's used?
* add a flag to enable the one-fixup-per-hunk behaviour?

some combination of the above?

**Comment by tummychow** [comment]
it's been a while since i sat down and stared at it, but i'm pretty sure the feature has a correctness issue (https://github.com/tummychow/git-absorb/pull/123 / https://github.com/tummychow/git-absorb/issues/116) so turning it on by default would be unwise
